tools/output/share/man/man3/GD.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.3
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GD 3"
.TH GD 3 "2014-04-08" "perl v5.8.7" "User Contributed Perl Documentation"
.SH "NAME"
GD.pm \- Interface to Gd Graphics Library
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    use GD;
.Ve
.PP
.Vb 2
\&    # create a new image
\&    $im = new GD::Image(100,100);
.Ve
.PP
.Vb 5
\&    # allocate some colors
\&    $white = $im->colorAllocate(255,255,255);
\&    $black = $im->colorAllocate(0,0,0);       
\&    $red = $im->colorAllocate(255,0,0);      
\&    $blue = $im->colorAllocate(0,0,255);
.Ve
.PP
.Vb 3
\&    # make the background transparent and interlaced
\&    $im->transparent($white);
\&    $im->interlaced('true');
.Ve
.PP
.Vb 2
\&    # Put a black frame around the picture
\&    $im->rectangle(0,0,99,99,$black);
.Ve
.PP
.Vb 2
\&    # Draw a blue oval
\&    $im->arc(50,50,95,75,0,360,$blue);
.Ve
.PP
.Vb 2
\&    # And fill it with red
\&    $im->fill(50,50,$red);
.Ve
.PP
.Vb 2
\&    # make sure we are writing to a binary stream
\&    binmode STDOUT;
.Ve
.PP
.Vb 2
\&    # Convert the image to GIF and print it on standard output
\&    print $im->gif;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fB\s-1GD\s0.pm\fR is a port of Thomas Boutell's gd graphics library (see
below).  \s-1GD\s0 allows you to create color drawings using a large number of
graphics primitives, and emit the drawings as \s-1GIF\s0 files.
.PP
\&\s-1GD\s0 defines the following three classes:
.ie n .IP """GD::Image""" 5
.el .IP "\f(CWGD::Image\fR" 5
.IX Item "GD::Image"
An image class, which holds the image data and accepts graphic
primitive method calls.
.ie n .IP """GD::Font""" 5
.el .IP "\f(CWGD::Font\fR" 5
.IX Item "GD::Font"
A font class, which holds static font information and used for text
rendering.
.ie n .IP """GD::Polygon""" 5
.el .IP "\f(CWGD::Polygon\fR" 5
.IX Item "GD::Polygon"
A simple polygon object, used for storing lists of vertices prior to
rendering a polygon into an image.
.PP
A Simple Example:
.PP
.Vb 1
\&        #!/usr/local/bin/perl
.Ve
.PP
.Vb 1
\&        use GD;
.Ve
.PP
.Vb 2
\&        # create a new image
\&        $im = new GD::Image(100,100);
.Ve
.PP
.Vb 5
\&        # allocate some colors
\&        $white = $im->colorAllocate(255,255,255);
\&        $black = $im->colorAllocate(0,0,0);       
\&        $red = $im->colorAllocate(255,0,0);      
\&        $blue = $im->colorAllocate(0,0,255);
.Ve
.PP
.Vb 3
\&        # make the background transparent and interlaced
\&        $im->transparent($white);
\&        $im->interlaced('true');
.Ve
.PP
.Vb 2
\&        # Put a black frame around the picture
\&        $im->rectangle(0,0,99,99,$black);
.Ve
.PP
.Vb 2
\&        # Draw a blue oval
\&        $im->arc(50,50,95,75,0,360,$blue);
.Ve
.PP
.Vb 2
\&        # And fill it with red
\&        $im->fill(50,50,$red);
.Ve
.PP
.Vb 2
\&        # make sure we are writing to a binary stream
\&        binmode STDOUT;
.Ve
.PP
.Vb 2
\&        # Convert the image to GIF and print it on standard output
\&        print $im->gif;
.Ve
.PP
Notes:
.IP "1. To create a new, empty image, send a \fInew()\fR message to GD::Image, passing it the width and height of the image you want to create.  An image object will be returned.  Other class methods allow you to initialize an image from a preexisting \s-1GIF\s0, \s-1GD\s0 or \s-1XBM\s0 file." 5
.IX Item "1. To create a new, empty image, send a new() message to GD::Image, passing it the width and height of the image you want to create.  An image object will be returned.  Other class methods allow you to initialize an image from a preexisting GIF, GD or XBM file."
.PD 0
.IP "2. Next you will ordinarily add colors to the image's color table. colors are added using a \fIcolorAllocate()\fR method call.  The three parameters in each call are the red, green and blue (rgb) triples for the desired color.  The method returns the index of that color in the image's color table.  You should store these indexes for later use." 5
.IX Item "2. Next you will ordinarily add colors to the image's color table. colors are added using a colorAllocate() method call.  The three parameters in each call are the red, green and blue (rgb) triples for the desired color.  The method returns the index of that color in the image's color table.  You should store these indexes for later use."
.IP "3. Now you can do some drawing!  The various graphics primitives are described below.  In this example, we do some text drawing, create an oval, and create and draw a polygon." 5
.IX Item "3. Now you can do some drawing!  The various graphics primitives are described below.  In this example, we do some text drawing, create an oval, and create and draw a polygon."
.IP "4. Polygons are created with a \fInew()\fR message to GD::Polygon.  You can add points to the returned polygon one at a time using the \fIaddPt()\fR method. The polygon can then be passed to an image for rendering." 5
.IX Item "4. Polygons are created with a new() message to GD::Polygon.  You can add points to the returned polygon one at a time using the addPt() method. The polygon can then be passed to an image for rendering."
.ie n .IP "5. When you're done drawing, you can convert the image into \s-1GIF\s0 format by sending it a \fIgif()\fR message.  It will return a (potentially large) scalar value containing the binary data for the image.  Ordinarily you will print it out at this point or write it to a file.  To ensure portability to platforms that differentiate between text and binary files, be sure to call ""binmode()"" on the file you are writing the image to." 5
.el .IP "5. When you're done drawing, you can convert the image into \s-1GIF\s0 format by sending it a \fIgif()\fR message.  It will return a (potentially large) scalar value containing the binary data for the image.  Ordinarily you will print it out at this point or write it to a file.  To ensure portability to platforms that differentiate between text and binary files, be sure to call \f(CWbinmode()\fR on the file you are writing the image to." 5
.IX Item "5. When you're done drawing, you can convert the image into GIF format by sending it a gif() message.  It will return a (potentially large) scalar value containing the binary data for the image.  Ordinarily you will print it out at this point or write it to a file.  To ensure portability to platforms that differentiate between text and binary files, be sure to call binmode() on the file you are writing the image to."
.PD
.SH "Method Calls"
.IX Header "Method Calls"
.Sh "Creating and Saving Images"
.IX Subsection "Creating and Saving Images"
.ie n .IP """new""" 5
.el .IP "\f(CWnew\fR" 5
.IX Item "new"
\&\f(CW\*(C`GD::Image::new(width,height)\*(C'\fR \fIclass method\fR
.Sp
To create a new, blank image, send a \fInew()\fR message to the GD::Image
class.  For example:
.Sp
.Vb 1
\&        $myImage = new GD::Image(100,100) || die;
.Ve
.Sp
This will create an image that is 100 x 100 pixels wide.  If you don't
specify the dimensions, a default of 64 x 64 will be chosen. If
something goes wrong (e.g. insufficient memory), this call will
return undef.
.ie n .IP """newFromGif""" 5
.el .IP "\f(CWnewFromGif\fR" 5
.IX Item "newFromGif"
\&\f(CW\*(C`GD::Image::newFromGif(FILEHANDLE)\*(C'\fR \fIclass method\fR
.Sp
This will create an image from a \s-1GIF\s0 file read in through the provided
filehandle.  The filehandle must previously have been opened on a
valid \s-1GIF\s0 file or pipe.  If successful, this call will return an
initialized image which you can then manipulate as you please.  If it
fails, which usually happens if the thing at the other end of the
filehandle is not a valid \s-1GIF\s0 file, the call returns undef.  Notice
that the call doesn't automatically close the filehandle for you.
But it does call \f(CW\*(C`binmode(FILEHANDLE)\*(C'\fR for you, on platforms where
this matters.
.Sp
To get information about the size and color usage of the information,
you can call the image query methods described below.
.Sp
.Vb 1
\&        Example usage:
.Ve
.Sp
.Vb 3
\&        open (GIF,"barnswallow.gif") || die;
\&        $myImage = newFromGif GD::Image(GIF) || die;
\&        close GIF;
.Ve
.ie n .IP """newFromXbm""" 5
.el .IP "\f(CWnewFromXbm\fR" 5
.IX Item "newFromXbm"
\&\f(CW\*(C`GD::Image::newFromXbm(FILEHANDLE)\*(C'\fR \fIclass method\fR
.Sp
This works in exactly the same way as \f(CW\*(C`newFromGif\*(C'\fR, but reads the
contents of an X Bitmap file:
.Sp
.Vb 3
\&        open (XBM,"coredump.xbm") || die;
\&        $myImage = newFromXbm GD::Image(XBM) || die;
\&        close XBM;
.Ve
.Sp
Note that this function also calls \f(CW\*(C`binmode(FILEHANDLE)\*(C'\fR before
reading from the filehandle.
.ie n .IP """newFromGd""" 5
.el .IP "\f(CWnewFromGd\fR" 5
.IX Item "newFromGd"
\&\f(CW\*(C`GD::Image::newFromGd(FILEHANDLE)\*(C'\fR \fIclass method\fR
.Sp
This works in exactly the same way as \f(CW\*(C`newFromGif\*(C'\fR, but reads the
contents of a \s-1GD\s0 file.  \s-1GD\s0 is Tom Boutell's disk-based storage format,
intended for the rare case when you need to read and write the image
to disk quickly.  It's not intended for regular use, because, unlike
\&\s-1GIF\s0 or \s-1JPEG\s0, no image compression is performed and these files can
become \fB\s-1BIG\s0\fR.
.Sp
.Vb 3
\&        open (GDF,"godzilla.gd") || die;
\&        $myImage = newFromGd GD::Image(GDF) || die;
\&        close GDF;
.Ve
.Sp
Note that this function also calls \f(CW\*(C`binmode(FILEHANDLE)\*(C'\fR before
reading from the supplied filehandle.
.ie n .IP """gif""" 5
.el .IP "\f(CWgif\fR" 5
.IX Item "gif"
\&\f(CW\*(C`GD::Image::gif\*(C'\fR \fIobject method\fR
.Sp
This returns the image data in \s-1GIF\s0 format.  You can then print it,
pipe it to a display program, or write it to a file.  Example:
.Sp
.Vb 5
\&        $gif_data = $myImage->gif;
\&        open (DISPLAY,"| display -") || die;
\&        binmode DISPLAY;
\&        print DISPLAY $gif_data;
\&        close DISPLAY;
.Ve
.Sp
Note the use of \f(CW\*(C`binmode()\*(C'\fR.  This is crucial for portability to
DOSish platforms.
.ie n .IP """gd""" 5
.el .IP "\f(CWgd\fR" 5
.IX Item "gd"
\&\f(CW\*(C`GD::Image::gd\*(C'\fR \fIobject method\fR
.Sp
This returns the image data in \s-1GD\s0 format.  You can then print it,
pipe it to a display program, or write it to a file.  Example:
.Sp
.Vb 2
\&        binmode MYOUTFILE;
\&        print MYOUTFILE $myImage->gd;
.Ve
.Sh "Color Control"
.IX Subsection "Color Control"
.ie n .IP """colorAllocate""" 5
.el .IP "\f(CWcolorAllocate\fR" 5
.IX Item "colorAllocate"
\&\f(CW\*(C`GD::Image::colorAllocate(red,green,blue)\*(C'\fR \fIobject method\fR
.Sp
This allocates a color with the specified red, green and blue
components and returns its index in the color table, if specified.
The first color allocated in this way becomes the image's background
color.  (255,255,255) is white (all pixels on).  (0,0,0) is black (all
pixels off).  (255,0,0) is fully saturated red.  (127,127,127) is 50%
gray.  You can find plenty of examples in /usr/X11/lib/X11/rgb.txt.
.Sp
If no colors are allocated, then this function returns \-1.
.Sp
Example:
.Sp
.Vb 3
\&        $white = $myImage->colorAllocate(0,0,0); #background color
\&        $black = $myImage->colorAllocate(255,255,255);
\&        $peachpuff = $myImage->colorAllocate(255,218,185);
.Ve
.ie n .IP """colorDeallocate""" 5
.el .IP "\f(CWcolorDeallocate\fR" 5
.IX Item "colorDeallocate"
\&\f(CW\*(C`GD::Image::colorDeallocate(colorIndex)\*(C'\fR \fIobject method\fR 
.Sp
This marks the color at the specified index as being ripe for
reallocation.  The next time colorAllocate is used, this entry will be
replaced.  You can call this method several times to deallocate
multiple colors.  There's no function result from this call.
.Sp
Example:
.Sp
.Vb 2
\&        $myImage->colorDeallocate($peachpuff);
\&        $peachy = $myImage->colorAllocate(255,210,185);
.Ve
.ie n .IP """colorClosest""" 5
.el .IP "\f(CWcolorClosest\fR" 5
.IX Item "colorClosest"
\&\f(CW\*(C`GD::Image::colorClosest(red,green,blue)\*(C'\fR \fIobject method\fR
.Sp
This returns the index of the color closest in the color table to the
red green and blue components specified.  If no colors have yet been
allocated, then this call returns \-1.
.Sp
Example:
.Sp
.Vb 1
\&        $apricot = $myImage->colorClosest(255,200,180);
.Ve
.ie n .IP """colorExact""" 5
.el .IP "\f(CWcolorExact\fR" 5
.IX Item "colorExact"
\&\f(CW\*(C`GD::Image::colorExact(red,green,blue)\*(C'\fR \fIobject method\fR
.Sp
This returns the index of a color that exactly matches the specified
red green and blue components.  If such a color is not in the color
table, this call returns \-1.
.Sp
.Vb 2
\&        $rosey = $myImage->colorExact(255,100,80);
\&        warn "Everything's coming up roses.\en" if $rosey >= 0;
.Ve
.ie n .IP """colorsTotal""" 5
.el .IP "\f(CWcolorsTotal\fR" 5
.IX Item "colorsTotal"
\&\f(CW\*(C`GD::Image::colorsTotal)\*(C'\fR \fIobject method\fR
.Sp
This returns the total number of colors allocated in the object.
.Sp
.Vb 1
\&        $maxColors = $myImage->colorsTotal;
.Ve
.ie n .IP """getPixel""" 5
.el .IP "\f(CWgetPixel\fR" 5
.IX Item "getPixel"
\&\f(CW\*(C`GD::Image::getPixel(x,y)\*(C'\fR \fIobject method\fR
.Sp
This returns the color table index underneath the specified
point.  It can be combined with \fIrgb()\fR
to obtain the rgb color underneath the pixel.
.Sp
Example:
.Sp
.Vb 2
\&        $index = $myImage->getPixel(20,100);
\&        ($r,$g,$b) = $myImage->rgb($index);
.Ve
.ie n .IP """rgb""" 5
.el .IP "\f(CWrgb\fR" 5
.IX Item "rgb"
\&\f(CW\*(C`GD::Image::rgb(colorIndex)\*(C'\fR \fIobject method\fR
.Sp
This returns a list containing the red, green and blue components of
the specified color index.
.Sp
Example:
.Sp
.Vb 1
\&        @RGB = $myImage->rgb($peachy);
.Ve
.ie n .IP """transparent""" 5
.el .IP "\f(CWtransparent\fR" 5
.IX Item "transparent"
\&\f(CW\*(C`GD::Image::transparent(colorIndex)\*(C'\fR \fIobject method\fR
.Sp
This marks the color at the specified index as being transparent.
Portions of the image drawn in this color will be invisible.  This is
useful for creating paintbrushes of odd shapes, as well as for
making \s-1GIF\s0 backgrounds transparent for displaying on the Web.  Only
one color can be transparent at any time. To disable transparency, 
specify \-1 for the index.  
.Sp
If you call this method without any parameters, it will return the
current index of the transparent color, or \-1 if none.
.Sp
Example:
.Sp
.Vb 6
\&        open(GIF,"test.gif");
\&        $im = newFromGif GD::Image(GIF);
\&        $white = $im->colorClosest(255,255,255); # find white
\&        $im->transparent($white);
\&        binmode STDOUT;
\&        print $im->gif;
.Ve
.Sh "Special Colors"
.IX Subsection "Special Colors"
\&\s-1GD\s0 implements a number of special colors that can be used to achieve
special effects.  They are constants defined in the \s-1GD::\s0
namespace, but automatically exported into your namespace when the \s-1GD\s0
module is loaded.
.ie n .IP """setBrush""" 5
.el .IP "\f(CWsetBrush\fR" 5
.IX Item "setBrush"
.PD 0
.ie n .IP """gdBrushed""" 5
.el .IP "\f(CWgdBrushed\fR" 5
.IX Item "gdBrushed"
.PD
\&\f(CW\*(C`GD::Image::setBrush( )\*(C'\fR and \f(CW\*(C`GD::gdBrushed\*(C'\fR
.Sp
You can draw lines and shapes using a brush pattern.  Brushes are 
just images that you can create and manipulate in the usual way. When
you draw with them, their contents are used for the color and shape of
the lines.
.Sp
To make a brushed line, you must create or load the brush first, then
assign it to the image using \f(CW\*(C`setBrush\*(C'\fR.  You can then draw in that
with that brush using the \f(CW\*(C`gdBrushed\*(C'\fR special color.  It's often 
useful to set the background of the brush to transparent so that 
the non-colored parts don't overwrite other parts of your image.
.Sp
Example:
.Sp
.Vb 6
\&        # Create a brush at an angle
\&        $diagonal_brush = new GD::Image(5,5);
\&        $white = $diagonal_brush->allocateColor(255,255,255);
\&        $black = $diagonal_brush->allocateColor(0,0,0);
\&        $diagonal_brush->transparent($white);
\&        $diagonal_brush->line(0,4,4,0,$black); # NE diagonal
.Ve
.Sp
.Vb 2
\&        # Set the brush
\&        $myImage->setBrush($diagonal_brush);
.Ve
.Sp
.Vb 2
\&        # Draw a circle using the brush
\&        $myImage->arc(50,50,25,25,0,360,gdBrushed);
.Ve
.ie n .IP """setStyle""" 5
.el .IP "\f(CWsetStyle\fR" 5
.IX Item "setStyle"
.PD 0
.ie n .IP """gdStyled""" 5
.el .IP "\f(CWgdStyled\fR" 5
.IX Item "gdStyled"
.PD
\&\f(CW\*(C`GD::Image::setStyle(@colors)\*(C'\fR and \f(CW\*(C`GD::gdStyled\*(C'\fR
.Sp
Styled lines consist of an arbitrary series of repeated colors and are
useful for generating dotted and dashed lines.  To create a styled
line, use \f(CW\*(C`setStyle\*(C'\fR to specify a repeating series of colors.  It
accepts an array consisting of one or more color indexes.  Then
draw using the \f(CW\*(C`gdStyled\*(C'\fR special color.  Another special color,
\&\f(CW\*(C`gdTransparent\*(C'\fR can be used to introduce holes in the line, as the
example shows.
.Sp
Example:
.Sp
.Vb 6
\&        # Set a style consisting of 4 pixels of yellow,
\&        # 4 pixels of blue, and a 2 pixel gap
\&        $myImage->setStyle($yellow,$yellow,$yellow,$yellow,
\&                           $blue,$blue,$blue,$blue,
\&                           gdTransparent,gdTransparent);
\&        $myImage->arc(50,50,25,25,0,360,gdStyled);
.Ve
.Sp
To combine the \f(CW\*(C`gdStyled\*(C'\fR and \f(CW\*(C`gdBrushed\*(C'\fR behaviors, you can specify
\&\f(CW\*(C`gdStyledBrushed\*(C'\fR.  In this case, a pixel from the current brush
pattern is rendered wherever the color specified in \fIsetStyle()\fR is
neither gdTransparent nor 0.
.ie n .IP """gdTiled""" 5
.el .IP "\f(CWgdTiled\fR" 5
.IX Item "gdTiled"
Draw filled shapes and flood fills using a pattern.  The pattern is
just another image.  The image will be tiled multiple times in order
to fill the required space, creating wallpaper effects.  You must call
\&\f(CW\*(C`setTile\*(C'\fR in order to define the particular tile pattern you'll use
for drawing when you specify the gdTiled color.
details.
.ie n .IP """gdStyled""" 5
.el .IP "\f(CWgdStyled\fR" 5
.IX Item "gdStyled"
The gdStyled color is used for creating dashed and dotted lines.  A
styled line can contain any series of colors and is created using the
\&\f(CW\*(C`setStyled\*(C'\fR command.
.Sh "Drawing Commands"
.IX Subsection "Drawing Commands"
.ie n .IP """setPixel""" 5
.el .IP "\f(CWsetPixel\fR" 5
.IX Item "setPixel"
\&\f(CW\*(C`GD::Image::setPixel(x,y,color)\*(C'\fR \fIobject method\fR 
.Sp
This sets the pixel at (x,y) to the specified color index.  No value
is returned from this method.  The coordinate system starts at the
upper left at (0,0) and gets larger as you go down and to the right.
You can use a real color, or one of the special colors gdBrushed, 
gdStyled and gdStyledBrushed can be specified.
.Sp
Example:
.Sp
.Vb 2
\&        # This assumes $peach already allocated
\&        $myImage->setPixel(50,50,$peach);
.Ve
.ie n .IP """line""" 5
.el .IP "\f(CWline\fR" 5
.IX Item "line"
\&\f(CW\*(C`GD::Image::line(x1,y1,x2,y2,color)\*(C'\fR \fIobject method\fR
.Sp
This draws a line from (x1,y1) to (x2,y2) of the specified color.  You
can use a real color, or one of the special colors gdBrushed, 
gdStyled and gdStyledBrushed.
.Sp
Example:
.Sp
.Vb 3
\&        # Draw a diagonal line using the currently defind
\&        # paintbrush pattern.
\&        $myImage->line(0,0,150,150,gdBrushed);
.Ve
.ie n .IP """dashedLine""" 5
.el .IP "\f(CWdashedLine\fR" 5
.IX Item "dashedLine"
\&\f(CW\*(C`GD::Image::dashedLine(x1,y1,x2,y2,color)\*(C'\fR \fIobject method\fR
.Sp
This draws a dashed line from (x1,y1) to (x2,y2) in the specified
color.  A more powerful way to generate arbitrary dashed and dotted
lines is to use the \fIsetStyle()\fR method described below and to draw with
the special color gdStyled.
.Sp
Example:
.Sp
.Vb 1
\&        $myImage->dashedLine(0,0,150,150,$blue);
.Ve
.ie n .IP """rectangle""" 5
.el .IP "\f(CWrectangle\fR" 5
.IX Item "rectangle"
\&\f(CW\*(C`GD::Image::rectangle(x1,y1,x2,y2,color)\*(C'\fR \fIobject method\fR
.Sp
This draws a rectangle with the specified color.  (x1,y1) and (x2,y2)
are the upper left and lower right corners respectively.  Both real 
color indexes and the special colors gdBrushed, gdStyled and 
gdStyledBrushed are accepted.
.Sp
Example:
.Sp
.Vb 1
\&        $myImage->rectangle(10,10,100,100,$rose);
.Ve
.ie n .IP """filledRectangle""" 5
.el .IP "\f(CWfilledRectangle\fR" 5
.IX Item "filledRectangle"
\&\f(CW\*(C`GD::Image::filledRectangle(x1,y1,x2,y2,color)\*(C'\fR \fIobject method\fR
.Sp
This draws a rectangle filed with the specified color.  You can use a
real color, or the special fill color gdTiled to fill the polygon
with a pattern.
.Sp
Example:
.Sp
.Vb 4
\&        # read in a fill pattern and set it
\&        open(GIF,"happyface.gif") || die;
\&        $tile = newFromGif GD::Image(GIF);
\&        $myImage->setTile($tile);
.Ve
.Sp
.Vb 2
\&        # draw the rectangle, filling it with the pattern
\&        $myImage->filledRectangle(10,10,150,200,gdTiled);
.Ve
.ie n .IP """polygon""" 5
.el .IP "\f(CWpolygon\fR" 5
.IX Item "polygon"
\&\f(CW\*(C`GD::Image::polygon(polygon,color)\*(C'\fR \fIobject method\fR 
.Sp
This draws a polygon with the specified color.  The polygon must be
created first (see below).  The polygon must have at least three
vertices.  If the last vertex doesn't close the polygon, the method
will close it for you.  Both real color indexes and the special 
colors gdBrushed, gdStyled and gdStyledBrushed can be specified.
.Sp
Example:
.Sp
.Vb 5
\&        $poly = new GD::Polygon;
\&        $poly->addPt(50,0);
\&        $poly->addPt(99,99);
\&        $poly->addPt(0,99);
\&        $myImage->polygon($poly,$blue);
.Ve
.ie n .IP """filledPolygon""" 5
.el .IP "\f(CWfilledPolygon\fR" 5
.IX Item "filledPolygon"
\&\f(CW\*(C`GD::Image::filledPolygon(poly,color)\*(C'\fR \fIobject method\fR
.Sp

tools/src/PDF-API2-0.41/blib/lib/PDF/API2/Resource/XObject/Image/GD.pm
#=======================================================================
#    ____  ____  _____              _    ____ ___   ____
#   |  _ \|  _ \|  ___|  _   _     / \  |  _ \_ _| |___ \
#   | |_) | | | | |_    (_) (_)   / _ \ | |_) | |    __) |
#   |  __/| |_| |  _|    _   _   / ___ \|  __/| |   / __/
#   |_|   |____/|_|     (_) (_) /_/   \_\_|  |___| |_____|
#
#   A Perl Module Chain to faciliate the Creation and Modification
#   of High-Quality "Portable Document Format (PDF)" Files.
#
#   Copyright 1999-2005 Alfred Reibenschuh <areibens@cpan.org>.
#
#=======================================================================
#
#   This library is free software; you can redistribute it and/or
#   modify it under the terms of the GNU Lesser General Public
#   License as published by the Free Software Foundation; either
#   version 2 of the License, or (at your option) any later version.
#
#   This library is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   Lesser General Public License for more details.
#
#   You should have received a copy of the GNU Lesser General Public
#   License along with this library; if not, write to the
#   Free Software Foundation, Inc., 59 Temple Place - Suite 330,
#   Boston, MA 02111-1307, USA.
#
#   $Id: GD.pm,v 1.8 2005/03/14 22:01:31 fredo Exp $
#
#=======================================================================
package PDF::API2::Resource::XObject::Image::GD;

BEGIN {

    use PDF::API2::Util;
    use PDF::API2::Basic::PDF::Utils;
    use PDF::API2::Resource::XObject::Image;

    use POSIX;

    use vars qw(@ISA $VERSION);

    @ISA = qw( PDF::API2::Resource::XObject::Image );

    ( $VERSION ) = '$Revision: 1.8 $' =~ /Revision: (\S+)\s/; # $Date: 2005/03/14 22:01:31 $

}
no warnings qw[ deprecated recursion uninitialized ];

=item $res = PDF::API2::Resource::XObject::Image::GD->new $pdf, $obj [, $name]

Returns a image object from a GD::Image.

=cut

sub new {
    my ($class,$pdf,$obj,$name,@opts) = @_;
    my $self;

    $class = ref $class if ref $class;

    $self=$class->SUPER::new($pdf,$name|| 'Jx'.pdfkey());
    $pdf->new_obj($self) unless($self->is_obj($pdf));

    $self->{' apipdf'}=$pdf;

    $self->read_gd($obj,@opts);

    return($self);
}

=item $res = PDF::API2::Resource::XObject::Image::GD->new_api $api, $obj [, $name]

Returns a image object. This method is different from 'new' that
it needs an PDF::API2-object rather than a Text::PDF::File-object.

=cut

sub new_api {
    my ($class,$api,@opts)=@_;

    my $obj=$class->new($api->{pdf},@opts);
    $obj->{' api'}=$api;

    return($obj);
}

sub read_gd {
    my $self = shift @_;
    my $gd = shift @_;
    my %opts = @_;
    
    my ($w,$h) = $gd->getBounds();
    my $c = $gd->colorsTotal();

    $self->width($w);
    $self->height($h);

    $self->bpc(8);
    $self->colorspace('DeviceRGB');

    if(UNIVERSAL::can($gd,'jpeg') && ($c > 256) && !$opts{-lossless}) {

        $self->filters('DCTDecode');
        $self->{' nofilt'}=1;
        $self->{' stream'}=$gd->jpeg(75);

    } elsif(UNIVERSAL::can($gd,'raw')) {

        $self->filters('FlateDecode');
        $self->{' stream'}=$gd->raw;

    } else {

        $self->filters('FlateDecode');
        for(my $y=0;$y<$h;$y++) {
            for(my $x=0;$x<$w;$x++) {
                my $index=$gd->getPixel($x,$y);
                my @rgb=$gd->rgb($index);
                $self->{' stream'}.=pack('CCC',@rgb);
            }
        }

    }

    return($self);
}



1;

__END__

=head1 AUTHOR

alfred reibenschuh

=head1 HISTORY

    $Log: GD.pm,v $
    Revision 1.8  2005/03/14 22:01:31  fredo
    upd 2005

    Revision 1.7  2004/12/16 00:30:55  fredo
    added no warn for recursion

    Revision 1.6  2004/06/15 09:14:54  fredo
    removed cr+lf

    Revision 1.5  2004/06/07 19:44:44  fredo
    cleaned out cr+lf for lf

    Revision 1.4  2004/05/28 11:29:01  fredo
    added -lossless param to gd images

    Revision 1.3  2003/12/08 13:06:10  Administrator
    corrected to proper licencing statement

    Revision 1.2  2003/11/30 17:37:16  Administrator
    merged into default

    Revision 1.1.1.1.2.2  2003/11/30 16:57:09  Administrator
    merged into default

    Revision 1.1.1.1.2.1  2003/11/30 16:00:42  Administrator
    added CVS id/log


=cut